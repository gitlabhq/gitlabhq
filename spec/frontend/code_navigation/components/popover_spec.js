import { shallowMount } from '@vue/test-utils';
import DocLine from '~/code_navigation/components/doc_line.vue';
import Popover from '~/code_navigation/components/popover.vue';
import Markdown from '~/vue_shared/components/markdown/non_gfm_markdown.vue';

const DEFINITION_PATH_PREFIX = 'http://gitlab.com';

const MOCK_CODE_DATA = Object.freeze({
  hover: [
    {
      language: 'javascript',
      tokens: [
        [
          {
            class: 'k',
            value: 'function',
          },
          {
            value: ' main() {',
          },
        ],
        [
          {
            value: '}',
          },
        ],
      ],
    },
  ],
  definition_path: 'test.js',
  definitionLineNumber: 20,
});

const MOCK_DOCS_DATA = Object.freeze({
  hover: [
    {
      language: null,
      value: '```console.log```',
    },
  ],
  definition_path: 'test.js#L20',
});

const MOCK_DATA_WITH_REFERENCES = Object.freeze({
  hover: [
    {
      language: null,
      value: 'console.log',
    },
  ],
  references: [{ path: 'index.js' }, { path: 'app.js' }],
  definition_path: 'test.js#L20',
});

let wrapper;

function factory({ position, data, definitionPathPrefix, blobPath = 'index.js' }) {
  wrapper = shallowMount(Popover, {
    propsData: { position, data, definitionPathPrefix, blobPath },
    stubs: { DocLine, Markdown },
  });
}

describe('Code navigation popover component', () => {
  it('renders popover', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_CODE_DATA,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('srender references tab with empty text when no references exist', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_CODE_DATA,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
    });

    expect(wrapper.find('[data-testid="references-tab"]').text()).toContain('No references found');
  });

  it('renders link with hash to current file', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_CODE_DATA,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
      blobPath: 'test.js',
    });

    expect(wrapper.find('[data-testid="go-to-definition-btn"]').attributes('href')).toBe('#L20');
  });

  it('renders list of references', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_DATA_WITH_REFERENCES,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
    });

    expect(wrapper.find('[data-testid="references-tab"]').exists()).toBe(true);
    expect(wrapper.findAll('[data-testid="reference-link"]')).toHaveLength(2);
  });

  describe('code output', () => {
    it('renders code output', () => {
      factory({
        position: { x: 0, y: 0, height: 0 },
        data: MOCK_CODE_DATA,
        definitionPathPrefix: DEFINITION_PATH_PREFIX,
      });

      expect(wrapper.findComponent({ ref: 'code-output' }).exists()).toBe(true);
      expect(wrapper.findComponent({ ref: 'doc-output' }).exists()).toBe(false);
    });
  });

  describe('documentation output', () => {
    it('renders code output', () => {
      factory({
        position: { x: 0, y: 0, height: 0 },
        data: MOCK_DOCS_DATA,
        definitionPathPrefix: DEFINITION_PATH_PREFIX,
      });

      expect(wrapper.findComponent({ ref: 'code-output' }).exists()).toBe(false);
      expect(wrapper.findComponent({ ref: 'doc-output' }).exists()).toBe(true);
    });

    it('renders markdown', () => {
      factory({
        position: { x: 0, y: 0, height: 0 },
        data: MOCK_DOCS_DATA,
        definitionPathPrefix: DEFINITION_PATH_PREFIX,
      });

      expect(wrapper.findComponent(Markdown).exists()).toBe(true);
    });
  });
});
