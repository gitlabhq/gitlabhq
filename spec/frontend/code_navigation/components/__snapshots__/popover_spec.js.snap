// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code navigation popover component renders popover 1`] = `
<div
  class="bs-popover-bottom code-navigation-popover gl-popover popover popover-font-size-normal show"
  style="left: 0px; top: 0px;"
>
  <div
    class="arrow"
    style="left: 0px;"
  />
  <gl-tabs-stub
    contentclass="gl-py-0"
    queryparamname="tab"
    value="0"
  >
    <gl-tab-stub
      title="Definition"
      titlelinkclass=""
    >
      <div
        class="code-navigation-popover-container overflow-auto"
      >
        <div>
          <pre
            class="bg-transparent border-0 code code-syntax-highlight-theme highlight m-0 text-wrap"
          >
            <span
              class="line"
              lang="javascript"
            >
              <span
                class="k"
              >
                function
              </span>
              <span>
                main() {
              </span>
              <br />
            </span>
            <span
              class="line"
              lang="javascript"
            >
              <span>
                }
              </span>
              <br />
            </span>
          </pre>
        </div>
      </div>
      <div
        class="border-top popover-body"
      >
        <gl-button-stub
          buttontextclasses=""
          category="primary"
          class="gl-w-full"
          data-testid="go-to-definition-btn"
          href="http://gitlab.com/test.js"
          icon=""
          size="medium"
          tag="button"
          target="_blank"
          type="button"
          variant="default"
        >
          Go to definition
        </gl-button-stub>
      </div>
    </gl-tab-stub>
    <gl-tab-stub
      class="py-2"
      data-testid="references-tab"
      titlelinkclass=""
    >
      <p
        class="gl-my-4 gl-px-4"
      >
        No references found
      </p>
    </gl-tab-stub>
  </gl-tabs-stub>
</div>
`;
